File: 00-每日一题/20200224_202.cpp
//
// Created by zanbo on 2020/2/24.
//

class Solution {

public:
    bool isHappy(int n) {
        unordered_set<int> mySet;
        while(1)
        {
            int num = newNum(n);
            if(num == 1)
                return true;
            if(mySet.find(num)!=mySet.end())
                return false;
            mySet.insert(num);
            n = num;
        }
    }

    int newNum(int x)
    {
        int sum=0;
        while(x)
        {
            sum += pow(x%10,2);
            x /= 10;
        }
        return sum;
    }
};

File: 00-每日一题/20200224_202.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        if n < 0:
            return False
        last_nums = set()

        while True:
            sum = 0
            while n:
                h = n % 10
                sum += h ** 2
                n = n // 10
            if sum == 1:
                return True
            if sum in last_nums:
                return False
            n = sum
            last_nums.add(n)

if __name__ == '__main__':

    obj = Solution()
    while True:

        n = int(input().strip())
        res = obj.isHappy(n)
        print(res)